Move requirements.txt to setup.py

PromyLOPh opened the pull request:
This *ensures* only known-good dependencies are used. By ignoring new, non-working versions we can concentrate on fixing the test suite. If we got that, we can come back later and selectively upgrade dependencies, one by one, fixing issues as they come up.

Edit: Uh, lemme check https://travis-ci.org/ArchiveTeam/wpull/jobs/504035201#L1437 before merging.

PromyLOPh committed:
Move requirements.txt to setup.py

Advantages:
- Single-command install (pip install .)
- Package can only be installed with known-good dependencies

PromyLOPh committed:
Move testing to setup.py as well

warcat and youtube-dl actually need to be installed, but tests_require
only downloads the egg and sets sys.path.

PromyLOPh committed:
Exclude broken html5lib/tornado versions

See https://github.com/ArchiveTeam/wpull/pull/402#issue-227999981 and
https://github.com/ArchiveTeam/wpull/pull/393#issuecomment-431101750

JustAnotherArchivist commented:
I know that this separation is often done on purpose since requirements.txt is for deployment while setup.py lists dependencies. Meaning the former should contain tested and confirmed-working versions but the latter only minimal requirements (e.g. excluding major versions with breaking changes). In this case, the versions of lxml and sqlalchemy at least are most likely not due to breaking changes.

PromyLOPh commented:
@JustAnotherArchivist We can look into lxml and sqlalchemy and remove the restriction when the tests pass and we’re confident it works. I still believe setup.py is the best place for dependency information. This enables the packet manager (pip) to make smart choices (or reject conflicts) and the user can simply run `pip install .`. Anyway, it’s up to you.

I’ve removed the commit changing dnspython3 to dnspython, >=1.13 breaks a test. (Well, doh.)

JustAnotherArchivist commented:
Yeah, I think I agree. I just wanted to mention why this separation might currently be the case.

Is there a good way to test different versions of dependencies, e.g. on Travis CI?

PromyLOPh commented:
Kind of: https://docs.travis-ci.com/user/languages/python/#testing-against-multiple-versions-of-dependencies-eg-django-or-flask

JustAnotherArchivist commented:
Yeah, I was hoping for a solution where we don't have to explicitly specify each version we want to test, but I guess that's a bit too much to ask.

PromyLOPh committed:
Upgrade dependencies

chardet: Works fine with 3.0.4. No API changes
namedlist: No new release since 1.7
sqlalchemy: Works fine with 1.3.1. NO UNIT TESTS!
yapsy: Version 1.12.0 works.

PromyLOPh committed:
setup.py: Document supported Python versions

Requires pip>=9 or setuptools>=24

PromyLOPh committed:
travis: Ignore required Python version

We want automated tests. Even for unsupported versions.
